homeassistant/components/ihc/__init__.py

```diff
@@ -128,38 +127,51 @@ def setup(hass: HomeAssistant, config: ConfigType) -> bool:
     return True
 
 
-def ihc_setup(hass, config, conf, controller_id):
+def ihc_setup(
+    hass: HomeAssistant,
+    config: ConfigType,
+    controller_conf: ConfigType,
+    controller_index: int,
+):
     """Set up the IHC integration."""
-    url = conf[CONF_URL]
-    username = conf[CONF_USERNAME]
-    password = conf[CONF_PASSWORD]
+    url = controller_conf[CONF_URL]
+    username = controller_conf[CONF_USERNAME]
+    password = controller_conf[CONF_PASSWORD]
 
     ihc_controller = IHCController(url, username, password)
     if not ihc_controller.authenticate():
         _LOGGER.error("Unable to authenticate on IHC controller")
         return False
-
-    if conf[CONF_AUTOSETUP] and not autosetup_ihc_products(
+    controller_id: str = ihc_controller.client.get_system_info()["serial_number"]
+    # Store controller configuration
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][controller_id] = {
+        IHC_CONTROLLER: ihc_controller,
+        CONF_INFO: controller_conf[CONF_INFO],
+        IHC_CONTROLLER_INDEX: controller_index,
+    }
+    if controller_conf[CONF_AUTOSETUP] and not autosetup_ihc_products(
         hass, config, ihc_controller, controller_id
     ):
         return False
-    # Manual configuration
-    get_manual_configuration(hass, config, conf, ihc_controller, controller_id)
-    # Store controller configuration
-    ihc_key = f"ihc{controller_id}"
-    hass.data[ihc_key] = {IHC_CONTROLLER: ihc_controller, IHC_INFO: conf[CONF_INFO]}
+    get_manual_configuration(hass, config, controller_conf, controller_id)
     # We only want to register the service functions once for the first controller
-    if controller_id == 0:
+    if controller_index == 0:
         setup_service_functions(hass)
     return True
 
 
-def get_manual_configuration(hass, config, conf, ihc_controller, controller_id):
+def get_manual_configuration(
+    hass: HomeAssistant,
+    config: ConfigType,
+    controller_conf,
+    controller_id: str,
+):
```

Comment:
Missing return value typing. Please type the whole signature when adding type annotations.

Reply:
Typings added for return value and all parameters. I also needed to change the controller_conf.get(platform) to controller_conf.get(platform,{}), to make mypy happy. Look like "for xx in None" does not give a runtime error - just does nothing as expected, but mypy don't like it.